### PR TITLE
Refactor parameter interpolation

### PR DESCRIPTION
**Pull request by mcwitt**
This PR unifies the alignment logic used for parameter interpolation, as suggested [here](https://github.com/proteneer/timemachine/pull/763#discussion_r907619290). This factors out a common pattern in the potential-specific alignment functions into a new helper, `align_idxs_and_params`.

The only change to the logic is in `align_chiral_bond_idxs_and_params`, which will now raise an error if there are duplicate (idxs, sign) pairs.

### CODE REVIEW

File: timemachine/fe/interpolate.py [hunk 1]
```diff
@@ -1,217 +1,120 @@
...
-
-    for idxs, k, sign in zip(dst_idxs, dst_params, dst_signs):
-        assert idxs[0] < idxs[-1]
-        key = (idxs, sign)
-        if key not in src_kv:
-            src_kv[key] = 0
-
-    return src_kv
+align_nonbonded_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: (0, 0, 0))
+align_chiral_atom_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: 0)
```
Comment (proteneer) [inline timemachine/fe/interpolate.py:96]: we may need to do one more check on the canonicalization of chiral terms (missing in master as well) for alignment purposes:

```
(0,1,2,3)
(0,3,2,1)
(0,2,1,3)
(2,1,3,0)
(3,1,0,2)
(1,3,2,0)
(3,0,2,1)
(1,2,0,3)
(2,0,1,3)
```

I'm not sure if we actually do this correctly!

Comment (proteneer) [reply timemachine/fe/interpolate.py:96]: nevermind - this should be taken care of by the canonicalization routine - will double check next week

Comment (proteneer) [reply timemachine/fe/interpolate.py:96]: [relevant bits](https://github.com/proteneer/timemachine/blob/master/timemachine/fe/single_topology_v3.py#L355)

Okay so the code currently doesn't quite do full canonicalization, but it's probably okay (and intended behavior). We define the chiral restraints as being "centered" at the `0th` index atom:

```
(0,1,2,3)
(0,3,2,1)
(0,2,1,3)
```
is distinct from 
```
(1,0,2,3)
(1,3,2,0)
(1,2,0,3)
```
(even though they result in the same sign). Closing this and sorry for the confusion.

**Comment by proteneer** [comment]
I'm okay with this change - I'd tag @maxentile  or @jkausrelay  for another pair of eyes though.

File: timemachine/fe/interpolate.py [hunk 2]
```diff
@@ -1,217 +1,138 @@
...
+    set
+        set of tuples (idxs, src_params, dst_params)
+
+    Examples
+    --------
+        >>> align_harmonic_angle_idxs_and_params = partial(align_idxs_and_params, make_default=lambda p: (0.0, p[1]))
+
+        For harmonic bonds, we align on idxs and fill missing
+        parameters with (0, d) where d is the equilibrium bond length
+
```
Comment (jkausrelay) [inline timemachine/fe/interpolate.py:58]: and 0 is the corresponding force constant.

File: timemachine/fe/interpolate.py [hunk 3]
```diff
@@ -1,217 +1,138 @@
...
-        key = (idxs, sign)
-        if key not in src_kv:
-            src_kv[key] = 0
-
-    return src_kv
+align_nonbonded_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: (0, 0, 0))
+align_chiral_atom_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: 0)
+align_torsion_idxs_and_params = partial(
+    align_idxs_and_params,
+    make_default=lambda p: (0, p[1], p[2]),
```
Comment (jkausrelay) [inline timemachine/fe/interpolate.py:117]: # p[1], p[2] is phase, period

File: timemachine/fe/interpolate.py [hunk 4]
```diff
@@ -1,217 +1,138 @@
...
+    key=lambda idxs, params: (idxs, params[2]),
```
Comment (jkausrelay) [inline timemachine/fe/interpolate.py:118]: `# use idxs and period as the key`

Comment (jkausrelay) [reply timemachine/fe/interpolate.py:118]: nit: params -> p for consistency

File: timemachine/fe/interpolate.py [hunk 5]
```diff
@@ -1,217 +1,138 @@
...
-
-    for key in src_kv.keys():
-        idxs, sign = key
-        src_k = src_kv[key]
-        dst_k = dst_kv[key]
-        res.add((idxs, sign, src_k, dst_k))
-
-    return res
+    return {
+        (idxs, s, p1, p2)
```
Comment (jkausrelay) [inline timemachine/fe/interpolate.py:125]: nit: s -> sign

File: timemachine/fe/interpolate.py [hunk 6]
```diff
@@ -1,217 +1,138 @@
...
+        for idxs, (s, p1), (_, p2) in align_idxs_and_params(
+            src_idxs,
+            zip(src_signs, src_params),
+            dst_idxs,
+            zip(dst_signs, dst_params),
+            make_default=lambda ps: (ps[0], 0),
```
Comment (jkausrelay) [inline timemachine/fe/interpolate.py:131]: nit: ps -> p for consistency with above
Also add a comment `# p[0] is the sign, 0 is the force constant`

**Comment by jkausrelay** [comment]
@mcwitt lgtm, few minor comments.

**Review by jkausrelay — APPROVED** [review]

File: timemachine/fe/interpolate.py (hunk 2 above)
Comment (mcwitt) [reply timemachine/fe/interpolate.py:64]: Thanks, done in edcc4e3

File: timemachine/fe/interpolate.py (hunk 3 above)
Comment (mcwitt) [reply timemachine/fe/interpolate.py]: Done in 0651472

File: timemachine/fe/interpolate.py (hunk 4 above)
Comment (mcwitt) [reply timemachine/fe/interpolate.py]: Both done in 0651472

File: timemachine/fe/interpolate.py (hunk 5 above)
Comment (mcwitt) [reply timemachine/fe/interpolate.py]: Done in 0651472

File: timemachine/fe/interpolate.py (hunk 6 above)
Comment (mcwitt) [reply timemachine/fe/interpolate.py]: Done in 0651472

**Pull request merged**